osbuild/meta.py

```diff
@@ -585,3 +629,50 @@ def get_schema(self, klass, name=None, version="1") -> Schema:
         self._schemata[(klass, name, version)] = schema
 
         return schema
+
+    def list_runners(self, distro: Optional[str] = None) -> List[RunnerInfo]:
+        """List all available runner modules"""
+        if not self._runners:
+            path = os.path.join(self.path, "runners")
+            names = filter(lambda f: os.path.isfile(f"{path}/{f}"),
+                           os.listdir(path))
+            paths = map(lambda n: os.path.join(path, n), names)
+
+            runners_mapped = map(RunnerInfo.from_path, paths)
+            runners_sorted = sorted(runners_mapped, key=lambda r: r.distro)
+            runners_sorted = sorted(runners_sorted, key=lambda r: r.version)
+
+            self._runners = list(runners_sorted)
+
+        runners = self._runners[:]
+        if distro:
+            runners = list(filter(lambda r: r.distro == distro, runners))
```

Comment:
I know we're not the biggest fans of comprehensions but the copy, filtering, and return can be done: `return [r for r in self._runners if r.distro == distro]`

Reply:
`I know we're not the biggest fans of comprehensions`, I don't think that is true. In `meta.py` alone we have a bunch. My hunch here was that if the comprehension was doing filtering, we can spell it out with `filter`. Then again a few lines up we (I) used a list comprehension. 🤷 

It also send me down a bit of a rabbit whole, curious how the implementation differs:

```
In [1]: import dis

In [2]: def f1(lst, distro):
   ...:     return [l for l in lst if l == distro]
   ...:

In [3]: def f2(lst, distro):
   ...:     return list(filter(lambda l: l == distro, lst))
   ...:

In [4]: x = ["a1", "a2", "b1", "b2"]

In [5]: x = ["a", "a", "b", "b"]

In [7]: dis.dis(f1)
  2           0 LOAD_CLOSURE             0 (distro)
              2 BUILD_TUPLE              1
              4 LOAD_CONST               1 (<code object <listcomp> at 0x10b4e28c0, file "<ipython-input-2-941e4956f9f6>", line 2>)
              6 LOAD_CONST               2 ('f1.<locals>.<listcomp>')
              8 MAKE_FUNCTION            8 (closure)
             10 LOAD_FAST                0 (lst)
             12 GET_ITER
             14 CALL_FUNCTION            1
             16 RETURN_VALUE

Disassembly of <code object <listcomp> at 0x10b4e28c0, file "<ipython-input-2-941e4956f9f6>", line 2>:
  2           0 BUILD_LIST               0
              2 LOAD_FAST                0 (.0)
        >>    4 FOR_ITER                 8 (to 22)
              6 STORE_FAST               1 (l)
              8 LOAD_FAST                1 (l)
             10 LOAD_DEREF               0 (distro)
             12 COMPARE_OP               2 (==)
             14 POP_JUMP_IF_FALSE        2 (to 4)
             16 LOAD_FAST                1 (l)
             18 LIST_APPEND              2
             20 JUMP_ABSOLUTE            2 (to 4)
        >>   22 RETURN_VALUE

In [11]: dis.dis(f2)
  2           0 LOAD_GLOBAL              0 (list)
              2 LOAD_GLOBAL              1 (filter)
              4 LOAD_CLOSURE             0 (distro)
              6 BUILD_TUPLE              1
              8 LOAD_CONST               1 (<code object <lambda> at 0x10d672b80, file "<ipython-input-3-1d9baf9af2e4>", line 2>)
             10 LOAD_CONST               2 ('f2.<locals>.<lambda>')
             12 MAKE_FUNCTION            8 (closure)
             14 LOAD_FAST                0 (lst)
             16 CALL_FUNCTION            2
             18 CALL_FUNCTION            1
             20 RETURN_VALUE

Disassembly of <code object <lambda> at 0x10d672b80, file "<ipython-input-3-1d9baf9af2e4>", line 2>:
  2           0 LOAD_FAST                0 (l)
              2 LOAD_DEREF               0 (distro)
              4 COMPARE_OP               2 (==)
              6 RETURN_VALUE
```

So `filter` is implemented in [C](https://github.com/python/cpython/blob/bbcf42449e13c0b62f145cd49d12674ef3d5bf64/Python/bltinmodule.c#L574) but calling back for the lambda.

It also seems that `filter` and `map` almost got removed from Python3 ([email[(https://www.artima.com/weblogs/viewpost.jsp?thread=98196)), so I am inclined to say we strive for consistency in the future and just use list comprehensions.